src/simple_rtt.rs

```diff
@@ -0,0 +1,198 @@
+//! This module implements a minimal and non standard conforming HTTP 1.0
+//! round-tripper that works with the bitcoind RPC server. This can be used
+//! if minimal dependencies are a goal and synchronous communication is ok.
+
+use ::HttpRoundTripper;
+
+use http;
+
+use std;
+use std::io::{BufRead, BufReader, Cursor, Write};
+use std::net::{ToSocketAddrs, TcpStream};
+use std::time::{Instant, Duration};
+
+/// Simple bitcoind JSON RPC client that implements the necessary subset of HTTP
+#[derive(Copy, Clone, Debug)]
+pub struct Tripper {
+    default_port: u16,
+    timeout: Duration,
+}
+
+/// Builder for simple bitcoind `Tripper`s
+#[derive(Clone, Debug)]
+pub struct Builder {
+    tripper: Tripper,
+}
+
+impl Default for Tripper {
+    fn default() -> Self {
+        Tripper {
+            default_port: 8332,
+            timeout: Duration::from_secs(15),
+        }
+    }
+}
+
+
+impl Builder {
+    /// Construct new `Builder` with default configuration
+    pub fn new() -> Builder {
+        Builder {
+            tripper: Tripper::new(),
+        }
+    }
+
+    /// Sets the port that the tripper will connect to in case none was specified in the URL of the
+    /// request.
+    pub fn default_port(mut self, port: u16) -> Self {
+        self.tripper.default_port = port;
+        self
+    }
+
+    /// Sets the timeout after which requests will abort if they aren't finished
+    pub fn timeout(mut self, timeout: Duration) -> Self {
+        self.tripper.timeout = timeout;
+        self
+    }
+
+    /// Builds the final `Tripper`
+    pub fn build(self) -> Tripper {
+        self.tripper
+    }
+}
+
+impl Tripper {
+    /// Construct a new `Tripper` with default parameters
+    pub fn new() -> Self {
+        Tripper::default()
+    }
+
+    /// Returns a builder for `Tripper`
+    pub fn builder() -> Builder {
+        Builder::new()
+    }
+}
+
+/// Try to read a line from a buffered reader. If no line can be read till the deadline is reached
+/// return a timeout error.
+fn get_line<R: BufRead>(reader: &mut R, deadline: Instant) -> Result<String, Error> {
+    let mut line = String::new();
+    while deadline > Instant::now() {
+        match reader.read_line(&mut line) {
```

Comment:
Isn't `read_line` blocking? I have a deja vu. :D

Reply:
lol, that's right, we never did get to the bottom of this

Reply:
Sorry I totally forgot about that PR. Some months ago I apparently wrote an answer but did it as a review comment accidentally, so GH never published it :smile: 

> I don't think that works at all in the general case :see_no_evil: even with the underlying stream being nonblocking, the buffered reader would still block until a newline is hit. I never really completed that PR, the main use case was to detect core answering correctly but much to slow (idk why tbh), assuming it always sends complete lines at a time. It's stupid. Nowadays I'd probably just use `async` and timeouts. Or if not possible maybe that would require a custom buffered reader based on a nonblocking underlying stream? The `thread::yield` hack is rather horrible though imo.

> EDIT: I remember, the problem was that core was answering to slowly and I had to wait for lines to arrive (was returning `Ok(0)` a lot) so I built this helper fn, but apparently thought it would be clever to have an option to give up after some time. But it's super unreliable as it assumes Core to send full lines quickly and just take time between them.

Reply:
This is even more confusing :). It **does** work, even with a blocking stream and even with `BufRead` adding blocking on top of that blocking. And I didn't notice any slowness, though to be honest I didn't benchmark either.